File: sfe/sfe/crypto/SFECipher.java
package sfe.crypto;

import java.security.GeneralSecurityException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.spec.AlgorithmParameterSpec;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;

// based on the javax.crypto.Cipher API, but does not
// completely implement it (currently)

/**
 * Secure Hash/XOR based encryption
 */
public class SFECipher {
	public final static int ENCRYPT_MODE = Cipher.ENCRYPT_MODE;
	public final static int DECRYPT_MODE = Cipher.DECRYPT_MODE;
	
	public static SFECipher getInstance(String cipher) 
	throws GeneralSecurityException {
		String[] cmp = cipher.split("/");
		MessageDigest md = MessageDigest.getInstance(cmp[0]);
		if (cmp.length == 1 || cmp.length == 2) {
			return new SFECipher(md, true);
		} else if (cmp.length == 3) {
			boolean use_padding = !"NoPadding".equals(cmp[2]);
			return new SFECipher(md, use_padding);
		} else {
			throw new GeneralSecurityException("Bad cipher: " + cipher);
		}
	}
	
	private SFECipher(MessageDigest md, boolean use_padding) {
		this.md = md;
		this.use_padding = use_padding;
	}
	
	MessageDigest md;
	int mode = -1;
	SecretKey sk;
	boolean use_padding;

	public void init(int mode, SecretKey sk) throws InvalidKeyException {
		this.mode = mode;
		this.sk = sk;
		this.md.reset();
	}
	
	void  init(int mode, SecretKey sk, AlgorithmParameterSpec params) 
	throws InvalidKeyException, InvalidAlgorithmParameterException {
		if (!(params instanceof IvParameterSpec))
			throw new InvalidAlgorithmParameterException();
		init(mode, sk);
		md.update(((IvParameterSpec) params).getIV());	
	}
	
	public void setIV(byte[] iv) {
		md.reset();
		md.update(iv);
	}
	
	public byte[] doFinal(byte[] data) 
	throws IllegalBlockSizeException, BadPaddingException {
		SecretKey key = sk;
		sk = null;
		switch(mode) {
		case ENCRYPT_MODE:
			mode = -1;
			return use_padding ? encrypt(key, data) : deencrypt(key, data);
		case DECRYPT_MODE:
			mode = -1;
			return use_padding ? decrypt(key, data) : deencrypt(key, data);
		default:
			throw new IllegalStateException("Invalid mode: " + mode);
		}
	}
	
	// padding-free mode
	private byte[] deencrypt(SecretKey key, byte[] data) {
		byte[] xkey = md.digest(key.getEncoded());
		if (xkey.length < data.length) {
			throw new IllegalStateException("data too long");
		}
		byte[] ret = new byte[data.length];
		for (int i=0; i<ret.length; ++i) {
			ret[i] = (byte) (data[i] ^ xkey[i]);
		}
		return ret;
	}
	
	private byte[] encrypt(SecretKey key, byte[] data) {
		byte[] xkey = md.digest(key.getEncoded());
		if (xkey.length - 1 < data.length) {
			throw new IllegalStateException("data too long");
		}
		byte[] ret = new byte[xkey.length];
		for (int i=0; i<data.length; ++i) {
			ret[i] = (byte) (data[i] ^ xkey[i]);
		}
		for (int i=data.length; i<ret.length - 1; ++i) {
			ret[i] = xkey[i];
		}
		ret[ret.length - 1] = (byte) (xkey[ret.length - 1] ^ data.length); 
		return ret;
	}
	
	private byte[] decrypt(SecretKey key, byte[] data) throws BadPaddingException {
		byte[] xkey = md.digest(key.getEncoded());
		if (xkey.length != data.length) {
			throw new BadPaddingException("Invalid data length");
		}
		byte[] dec = new byte[data.length];
		for (int i=0; i<dec.length; ++i) {
			dec[i] = (byte) (data[i] ^ xkey[i]);
		}
		int len = dec[dec.length - 1];
		if (len > dec.length - 1 || len < 0)
			throw new BadPaddingException("Invalid data range");
		for (int i=len; i<dec.length-1; ++i) {
			if (dec[i] != 0)
				throw new BadPaddingException("Invalid data range");
		}
		byte[] ret = new byte[len];
		System.arraycopy(dec, 0, ret, 0, len);
		//System.out.println("len is " + len + "  xkey is " + xkey.length);
		return ret;
	}
}

File: sfe/C-sfe/src/crypto/cipher/SFECipher.cpp
/*
 * SFECipher.cpp
 *
 *  Created on: Sep 1, 2009
 *      Author: louis
 */

#include "SFECipher.h"

namespace crypto {
namespace cipher {


byte_buf SFECipher::md_xkey(const byte_buf &data) {
	byte_buf xkey(20);
	SHA1_Update(&ctx, &data[0], data.size());
	SHA1_Final(&xkey[0], &ctx);
	return silly_move(xkey);
}
// TODO: why do these take a key* ?
// padding-free mode
byte_buf SFECipher::deencrypt(const SFEKey *key, const byte_buf &data) {
	byte_buf xkey = md_xkey(*key->buf);
	if (xkey.size() < data.size()) {
		throw bad_argument("data too long");
	}

	byte_buf ret(data.size());
	for (uint i=0; i<ret.size(); ++i) {
		ret[i] = (data[i] ^ xkey[i]);
	}
	//std::cout << "de len is " << data.size() << "  xkey is " << xkey.size() << std::endl;
	return silly_move(ret);
}

byte_buf SFECipher::encrypt(const SFEKey *key, const byte_buf &data) {
	byte_buf xkey = md_xkey(*key->buf);
	if (xkey.size() - 1 < data.size()) {
		throw bad_argument("data too long");
	}
	byte_buf ret(xkey.size());
	for (uint i=0; i<data.size(); ++i) {
		ret[i] = (data[i] ^ xkey[i]);
	}
	for (uint i=data.size(); i<ret.size() - 1; ++i) {
		ret[i] = xkey[i];
	}
	ret[ret.size() - 1] = (xkey[ret.size() - 1] ^ data.size());
	return silly_move(ret);
}

byte_buf SFECipher::decrypt(const SFEKey *key, const byte_buf &data) {
	byte_buf xkey = md_xkey(*key->buf);
	if (xkey.size() != data.size()) {
		throw bad_padding("Invalid data length");
	}
	byte_buf dec(data.size());
	for (uint i=0; i<dec.size(); ++i) {
		dec[i] = (data[i] ^ xkey[i]);
	}
	uint len = dec[dec.size() - 1];
	if (len > dec.size() - 1)
		throw bad_padding("Invalid data range");
	for (uint i=len; i<dec.size()-1; ++i) {
		if (dec[i] != 0)
			throw bad_padding("Invalid data range");
	}
	dec.resize(len);
	//byte_buf ret = new byte[len];
	//System.arraycopy(dec, 0, ret, 0, len);
	//std::cout << "len is " << len << "  xkey is " << xkey.size() << std::endl;
	return silly_move(dec);
}



}
}

